tests/python/relax/test_inline_functions.py

```diff
@@ -0,0 +1,363 @@
+
+    When inlining, each instance of the inlined function may have a
+    different value for the symbolic variables it uses.
+    """
+
+    @I.ir_module
+    class Before:
+        @R.function(private=True)
+        def main(A: R.Tensor([16, 16]), B: R.Tensor([32, 32])):
+            A_out: R.Tensor([16, 16]) = Before.subroutine(A)
+            B_out: R.Tensor([32, 32]) = Before.subroutine(B)
+            return (A_out, B_out)
+
+        @R.function(private=True)
+        def subroutine(Input: R.Tensor(["n", "m"])) -> R.Tensor(["n", "m"]):
+            Output = Input + Input
+            return Output
+
+    @R.function(private=True)
+    def expected(A: R.Tensor([16, 16]), B: R.Tensor([32, 32])):
+        A_out: R.Tensor([16, 16]) = A + A
+        B_out: R.Tensor([32, 32]) = B + B
+        return (A_out, B_out)
+
+    after = Before["main"].inline_functions({"subroutine": Before["subroutine"]})
+    tvm.ir.assert_structural_equal(expected, after)
+
+
+def test_inline_nested_subroutine_calls():
+    """A private function may itself require inlining"""
+
+    @I.ir_module
+    class Before:
+        @R.function(private=True)
+        def main(A: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+            B = A * A
+            D = Before.subroutine(B)
+            E = D + D
+            return E
+
+        @R.function(private=True)
+        def subroutine(B: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+            C = R.concat([B, B], axis=1)
+            D = Before.subsubroutine(C)
+            return D
+
+        @R.function(private=True)
+        def subsubroutine(C: R.Tensor([16, 32], "int32")) -> R.Tensor([16, 32], "int32"):
+            D = C * C * C
+            return D
+
+    @R.function(private=True)
+    def expected(A: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+        B = A * A
+        C = R.concat([B, B], axis=1)
+        D = C * C * C
+        E = D + D
+        return E
+
+    after = Before["main"].inline_functions(
+        {
+            "subroutine": Before["subroutine"],
+            "subsubroutine": Before["subsubroutine"],
+        }
+    )
+    tvm.ir.assert_structural_equal(expected, after)
+
+
+def test_error_when_inlining_recursive_function():
+    """Inlining a recursive function call should raise an error"""
+
+    @I.ir_module
+    class Before:
+        @R.function(private=True)
+        def main():
+            B = Before.subroutine()
+            return B
+
+        @R.function(private=True)
+        def subroutine() -> R.Tensor([], "int64"):
+            R.func_attr({"relax.force_pure": True})
+            cond = R.call_packed("dummy_function", sinfo_args=R.Tensor([], "bool"))
+            if cond:
+                Out = Before.subroutine()
+            else:
+                Out = R.const(0, "int64")
+
+            return Out
+
+    with pytest.raises(Exception):
+        Before["main"].inline_functions({"subroutine": Before["subroutine"]})
+
```

Comment:
A couple of cases that are also worth testing:
1. Inlining within an inlined function
2. Making sure that mutually recursive functions can't be inlined either (since if you can inline within an inlined function, mutual recursion would also cause infinite inlining)

Reply:
Good catch.  I have test cases for theses two cases for the `InlinePrivateFunctions` transform, but not for the explicit user-driven calls.

Reply:
Good catch, thank you!  The inlining within an inlined function should already be tested in `test_inline_nested_subroutine_calls`.  The second case I had missed, and have now added `test_error_when_inlining_mutually_recursive_functions`.